client/client-multi/client-c/src/kaa/kaa_notification_manager.c

```diff
@@ -0,0 +1,1001 @@
+            KAA_LOG_TRACE(self->logger, KAA_ERR_NOMEM, "Failed to create storage for topic listeners.");
+            return KAA_ERR_NOMEM;
+        }
+    } else {
+        if (kaa_list_find_next(self->topics_listeners, &kaa_find_topic_listener_by_id, &id)) {
+            KAA_LOG_WARN(self->logger, KAA_ERR_ALREADY_EXISTS, "Failed to add topic the listener: the listener is already subscribed.");
+            return KAA_ERR_ALREADY_EXISTS;
+        }
+        wrapper = (kaa_topic_listener_wrapper_t *) KAA_MALLOC(sizeof(kaa_topic_listener_wrapper_t));
+        KAA_RETURN_IF_NIL(wrapper, KAA_ERR_NOMEM);
+        wrapper->listener = *listener;
+        if (!kaa_list_push_back(self->topics_listeners, wrapper)) {
+            KAA_LOG_ERROR(self->logger, KAA_ERR_NOMEM, "Failed to add the topic listener.");
+            return KAA_ERR_NOMEM;
+        }
+    }
+    *topic_listener_id = wrapper->id = id;// for user to have convenient way to address notification listener
+    return KAA_ERR_NONE;
+}
+
+kaa_error_t kaa_remove_topic_list_listener(kaa_notification_manager_t *self, uint32_t *topic_listener_id)
+{
+    KAA_RETURN_IF_NIL2(self, topic_listener_id, KAA_ERR_BADPARAM);
+    kaa_error_t err = kaa_list_remove_first(&self->topics_listeners, &kaa_find_topic_listener_by_id, topic_listener_id, &kaa_topic_listener_destroy);
+    if (err) {
+        KAA_LOG_WARN(self->logger, err, "Failed to remove the topic listener: the listener is not found");
+    }
+    if (!kaa_list_get_size(self->topics_listeners)) {
+        KAA_LOG_TRACE(self->logger, err, "Topic listeners list is empty now.");
+        self->topics_listeners = NULL;
+    }
+    return err;
+}
+
+kaa_error_t kaa_get_topics(kaa_notification_manager_t *self, kaa_list_t **topics)
+{
+    KAA_RETURN_IF_NIL2(self, topics, KAA_ERR_BADPARAM);
+    if (!self->topics) {
+        *topics = NULL;
+        KAA_LOG_WARN(self->logger, KAA_ERR_BAD_STATE, "Failed retrieve the topic list.The topic list is empty.");
+        return KAA_ERR_BAD_STATE;
+    }
+    *topics = self->topics;
+    return KAA_ERR_NONE;
+}
+
+static kaa_error_t kaa_notify_topic_update_subscribers(kaa_notification_manager_t *self, kaa_list_t *topics)
+{
+    KAA_RETURN_IF_NIL2(self, topics, KAA_ERR_BADPARAM);
+    kaa_list_t *current_listener_node = NULL;
+    if (!self->topics_listeners) {
+        KAA_LOG_TRACE(self->logger, KAA_ERR_NONE, "There is no topic listeners to be notified");
+        return KAA_ERR_NONE;
+    } else {
+        current_listener_node = self->topics_listeners;
+    }
+    while (current_listener_node) {
+        kaa_topic_listener_wrapper_t *wrapper = (kaa_topic_listener_wrapper_t *)kaa_list_get_data(current_listener_node);
+        if (wrapper->listener.callback) {
+            wrapper->listener.callback(wrapper->listener.context, topics);
+        } else {
+            KAA_LOG_WARN(self->logger, KAA_ERR_BADDATA, "Get NULL, expected callback, listener's id = %d.", wrapper->id);
```

Comment:
I think we have to guarantee non-null callback on the early stage, when the user adds it.